docs/compatibility.md

```diff
@@ -0,0 +1,183 @@
+# StableHLO Compatibility
+
+## Compatibility Guarantees
+
+(G1) libStablehlo provides 6 months of backward compatibility, i.e. StableHLO programs serialized by an old version of libStablehlo have the same semantics when deserialized by new versions of libStablehlo within the compatibility window.
+
+(G2) libStablehlo provides 3 weeks of forward compatibility, i.e. StableHLO programs serialized by a new version of libStablehlo have the same semantics when deserialized by old versions of libStablehlo within the compatibility window, unless they are using new features introduced since the old version.
+
+(G3) Source compatibility for C, C++ and Python APIs within libStablehlo is an aspirational goal. At the moment, we don't offer source compatibility guarantees, but please let us know if this is an important use case for you, and we can have a discussion about supporting it.
+
+\*StableHLO semantics is defined by the StableHLO specification and can be tested using the StableHLO interpreter. While we're bootstrapping the StableHLO specification, we will be referring to a combination of [XLA's Operation Semantics](https://www.tensorflow.org/xla/operation_semantics) and existing implementations to define the semantics.
+
+## StableHLO Programs
+
+For the purposes of compatibility guarantees, we define StableHLO programs as MLIR programs that include ops, attributes, and types specified below. This definition is based on our experience with lowering JAX/PyTorch/TensorFlow programs to MHLO and interviewing various groups using MHLO. Please let us know if you're interested in expanding this definition, and we can have a discussion about this.
+
+|  Dialect     | Ops, attributes and types |
+| ------------ | ----------- |
+| Arith        | `AddIOp, CmpIOp, CmpIPredicateAttr, ConstantOp, DivSIOp, ExtSIOp, IndexCastOp, MaxSIOp, MinSIOp, MulIOp, SelectOp, SubIOp, TruncIOp` |
```

Comment:
Technically speaking how is compatibility going to be achieved for ops in upstream dialects, if upstream decides to change them?

Reply:
More thought is still needed on handling upstream changes. The current plan only handles certain types of changes in upstream dialects. We are still iterating on how to handle a broader variety of changes to upstream ops / attributes / types.

Current plan: During serialization, convert the upstream op to a forked version of the op in StableHLO, during deserialization, convert it back. If StableHLO controls the serialization boundary, and we have upgrade/downgrade machinery, we can handle certain types of changes. This image in the proposal was an example of a Migration Change when `FuncOp` was moved from the builtin to func dialect.

![](https://raw.githubusercontent.com/openxla/stablehlo/68b5c357de519649dfef0561e4bb813bd0191aca/docs/images/dependent_dialect_change.png)

For instances where the semantics of an upstream op change, more exploration is still needed on our end.